src/libraries/System.Private.CoreLib/src/System/Array.cs

```diff
@@ -1871,6 +1872,19 @@ public static bool TrueForAll<T>(T[] array, Predicate<T> match)
             return true;
         }
 
+        /// <summary>
+        /// Gets the maximum number of elements that may be contained in an array of the specified type.
+        /// </summary>
+        /// <typeparam name="T">The type of array element.</typeparam>
+        /// <returns>The maximum count of elements allowed of array of the given type.</returns>
+        /// <remarks>
+        /// This methods returns runtime limitation.
+        /// The limitation is counted in elements, not bytes.
+        /// There's no guarantee that allocation under the size would succeed.
+        /// </remarks>
+        public static int GetMaxLength<T>() =>
```

Comment:
This needs `[MethodImpl(MethodImplOptions.AggressiveInlining)]`

Reply:
Why?

Reply:
@jkotas 

```csharp
static int GetMaxLength<T>() => Unsafe.SizeOf<T>() == 1 ? 1111 : 2222;

static int Foo() => GetMaxLength<Byte>();
```

codegen for `Foo()`:
```asm
; optimized code
G_M8231_IG01:
       50                   push     rax
G_M8231_IG02:
       E8B2FDFFFF           call     rut.Program:GetMaxLength():int
       90                   nop
G_M8231_IG03:
       4883C408             add      rsp, 8
       C3                   ret
```

Reason why it's not inlined:
```
Inline candidate callsite is boring.  Multiplier increased to 1.3.
calleeNativeSizeEstimate=227
callsiteNativeSizeEstimate=55
benefit multiplier=1.3
threshold=71
Native estimate for function size exceeds threshold for inlining 22.7 > 7.1 (multiplier = 1.3)


Inline expansion aborted, inline not profitable
```

Reply:
maybe something wrong with my local build (i tuned the inliner a bit) but sharplab shows the same output: https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aaGeHvf+PGsXJIGASwB2GBgHEYGALLYEAGRiSA5hgAWAHgAqAPgAUASgYBeI03IA6AEoBXaePww7LCPgAO4gDYwUADKQQBu4mAwuHYAqpK42ABmHsHiAF4wAPJJhqYWlpYM5AwA/MXklQwgDKR1pADcXLSiEtIMAGIQEOZWNvJKKupaunoAQgCeGDD5TdS8fNQAvkA===

Reply:
Hmmm, AggressiveInlining may help; but it is still going to have a problem with the generic inlining limitations. It won't get inlined when T is a shared generic type.

Reply:
Could `typeof(T).IsValueType` help? This should eliminate the `SizeOf` arm in shared generic of reference types.

Reply:
Sharplab shows that only `AggressiveInlining` is enough. I wonder if the behavior is "reliable".
Moreover, is there a way to test a method "should be inlined"?

Reply:
The problematic pattern is when the method is called from another generic method:

```csharp
class My
{    
    static int GetMaxLength<T>() => System.Runtime.CompilerServices.Unsafe.SizeOf<T>() == 1 ? 1111 : 2222;

    static int Foo<T>()
    {
        .... other code that makes Foo too big to be inlined ...
        GetMaxLength<T>();
    }
}
```

Reply:
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4a9BrIAW2KGIAy2YO258BG6poDMTUgyEMA3jRoMPDd5+8eA2gCyMBg6EAAmAJLikgAUQSHhUWKSAPJifBBcuCwAggDmebC4uLwKMBFckrxc1XkAlAC6tkzkSAzVGAwA4sEB2AiGMFx5IQA8ACoAfDF1vgwAvJMMAKpZ2ABmgrK8AF4wKesT03UL8wzkDAD85+S3DCAMpE+kVpqt7TwMAGIQEDMLSx6GD6AyGIx0owAQgBPDAwY5WTxeagAXzc1moSLmND8ACleBgelwZMoYhhoWIYBB1jEOnU6mgGPjCUMSWAyRSqTTmHR6U1aHYyI4jjRXJjPMw2h1vv9Fo4WECQYNhmMpjMrCigA===

`AggressiveInlining` isn't solving shared generic.

If I add `typeof(T).IsValueType`, the `int` version will not be inlined.

Reply:
So what's the desired approach for this method?
If `typeof(T)` is unused, JIT will be able to inline it.

Should I check latest JIT using disasmo?

Reply:
Interesting. Giving this code:
```csharp
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine(Dumped());
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static int GetMaxLength<T>()
            => Unsafe.SizeOf<T>() == 1 ? 1111 : 2222;

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static int GetConst<T>() => 1234;

        // [MethodImpl(MethodImplOptions.NoInlining)]
        public static int M<T>() => GetMaxLength<T>() + GetConst<T>();

        [MethodImpl(MethodImplOptions.NoInlining)]
        public static int Dumped() => M<string>();
    }
```

And use latest master branch, dumping with disasmo, `Dumped` will be a constant return now.

Could we view the JIT dump and performance of list related method?

Reply:
I tried JITDump on actual code, but seems that it doesn't dump generic method. The method I picked is `List<T>.EnsureCapacity`.

Reply:
@EgorBo It seems that current JIT can handle this. I want some instructions to prove it. Options could be:

1. Compare the JIT result. superpmi provides instruction for comparing different versions of jit, but no instruction for different versions of corelib.

2. Get the JIT dump of shared generic method.